tools/hdr/local_tone_map.cc

```diff
@@ -0,0 +1,378 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "lib/extras/codec.h"
+#include "lib/extras/tone_mapping.h"
+#include "lib/jxl/convolve.h"
+#include "lib/jxl/enc_color_management.h"
+#include "lib/jxl/enc_gamma_correct.h"
+#include "lib/jxl/image_bundle.h"
+#include "tools/args.h"
+#include "tools/cmdline.h"
+#include "tools/thread_pool_internal.h"
+
+namespace jxl {
+namespace {
+
+constexpr WeightsSeparable5 kPyramidFilter = {
+    {HWY_REP4(.375f), HWY_REP4(.25f), HWY_REP4(.0625f)},
+    {HWY_REP4(.375f), HWY_REP4(.25f), HWY_REP4(.0625f)}};
+
+// Expects sRGB input.
+// Will call consumer(x, y, contrast) for each pixel.
+template <typename Consumer>
+void Contrast(const jxl::Image3F& image, const Consumer& consumer,
+              ThreadPool* const pool) {
+  static constexpr WeightsSymmetric3 kLaplacianWeights = {
+      {HWY_REP4(-4)}, {HWY_REP4(1)}, {HWY_REP4(0)}};
+  ImageF grayscale(image.xsize(), image.ysize());
+  static constexpr float kLuminances[3] = {0.2126, 0.7152, 0.0722};
+  for (size_t y = 0; y < image.ysize(); ++y) {
+    const float* const JXL_RESTRICT input_rows[3] = {
+        image.PlaneRow(0, y), image.PlaneRow(1, y), image.PlaneRow(2, y)};
+    float* const JXL_RESTRICT row = grayscale.Row(y);
+
+    for (size_t x = 0; x < image.xsize(); ++x) {
+      row[x] = LinearToSrgb8Direct(
+          kLuminances[0] * Srgb8ToLinearDirect(input_rows[0][x]) +
+          kLuminances[1] * Srgb8ToLinearDirect(input_rows[1][x]) +
+          kLuminances[2] * Srgb8ToLinearDirect(input_rows[2][x]));
+    }
+  }
+
+  ImageF laplacian(image.xsize(), image.ysize());
+  Symmetric3(grayscale, Rect(grayscale), kLaplacianWeights, pool, &laplacian);
+  for (size_t y = 0; y < image.ysize(); ++y) {
+    const float* const JXL_RESTRICT row = laplacian.ConstRow(y);
+    for (size_t x = 0; x < image.xsize(); ++x) {
+      consumer(x, y, std::abs(row[x]));
+    }
+  }
+}
+
+template <typename Consumer>
+void Saturation(const jxl::Image3F& image, const Consumer& consumer) {
```

Comment:
stddev(r/g/b - average-rgb)?

This seems like a pretty primitive way to compute saturation - can it be improved?

Reply:
I would suggest something like the distance from origin in the Lab (a,b) plane or the XYB (scaledX,B-Y) plane, if that's not too expensive.

Reply:
sounds good as long as there is a todo :)